code/artificial_intelligence/src/a_star/astar.c

```diff
@@ -0,0 +1,253 @@
+                ++p_len;
+                path = (int*)realloc(path, p_len * sizeof(int));
+                path[p_len - 1] = current;
+            }
+        }
+ 
+        for (i = 0; i < o_len; i++) {
+            if (open[i] == current) {
+                if (i not_eq (o_len - 1)) {
+                    for (j = i; j < (o_len - 1); j++) {
+                        open[j] = open[j + 1];
+                    }
+                }
+                --o_len;
+                open = (int*)realloc(open, o_len * sizeof(int));
+                break;
+            }
+        }
+ 
+        ++c_len;
+        closed = (int*)realloc(closed, c_len * sizeof(int));
+        closed[c_len - 1] = current;
+ 
+        for (i = 0; i < stops[current].n_len; i++) {
+            b = 0;
+ 
+            for (j = 0; j < c_len; j++) {
+                if (routes[stops[current].n[i]].y == closed[j]) {
+                    b = 1;
+                }
+            }
+ 
+            if (b) {
+                continue;
+            }
+ 
+            tempg = stops[current].g + routes[stops[current].n[i]].d;
+ 
+            b = 1;
+ 
+            if (o_len > 0) {
+                for (j = 0; j < o_len; j++) {
+                    if (routes[stops[current].n[i]].y == open[j]) {
+                        b = 0;
+                    }
+                }
+            }
+ 
+            if (b or (tempg < stops[routes[stops[current].n[i]].y].g)) {
+                stops[routes[stops[current].n[i]].y].from = current;
+                stops[routes[stops[current].n[i]].y].g = tempg;
+                stops[routes[stops[current].n[i]].y].f = stops[routes[stops[current].n[i]].y].g + stops[routes[stops[current].n[i]].y].h;
+ 
+                if (b) {
+                    ++o_len;
+                    open = (int*)realloc(open, o_len * sizeof(int));
+                    open[o_len - 1] = routes[stops[current].n[i]].y;
+                }
+            }
+        }
+    }
+ 
+    for (i = 0; i < map_size_rows; i++) {
+        for (j = 0; j < map_size_cols; j++) {
+            if (map[i][j]) {
+                putchar(0xdb);
+            } else {
+                b = 0;
+                for (k = 0; k < p_len; k++) {
+                    if (ind[i][j] == path[k]) {
+                        ++b;
+                    }
+                }
+                if (b) {
+                    putchar('x');
+                } else {
+                    putchar('.');
+                }
+            }
+        }
+        putchar('\n');
+    }
+ 
+    if (not found) {
+        puts("IMPOSSIBLE");
+    } else {
+        printf("path cost is %d:\n", p_len);
+        for (i = p_len - 1; i >= 0; i--) {
+            printf("(%1.0f, %1.0f)\n", stops[path[i]].col, stops[path[i]].row);
+        }
+    }
+ 
+    for (i = 0; i < s_len; ++i) {
+        free(stops[i].n);
+    }
+    
+    free(stops);
+    free(routes);
+    free(path);
+    free(open);
+    free(closed);
+ 
+    return (0);
+}
+ 
```

Comment:
Only one line at the end with a newline. Last line has some space or something in it.